Added workflow

soumyaDghosh opened the pull request:
Also adds support for dev releases. Adds `-dev` after the version if it's a dev version. Also, doesn't creates a call-for-testing issue for dev versions and releases dev versions to `edge` channel instead of candidate.

soumyaDghosh committed:
update to latest release and add workflow

Also adds support for dev releases. Adds `-dev` after the version if it's a dev version. Also, doesn't creates a call-for-testing issue for dev versions and releases dev versions to `edge` channel instead of candidate.

soumyaDghosh committed:
change branch name

soumyaDghosh commented:
Branch renaming is needed. I was also thinking to move to `core22`, but regarding classic snaps, I am currently unsure, if they'll work everywhere

popey commented:
Please do *not* move to core22. You will break the application.

soumyaDghosh commented:
No I didn't it's still on core18

popey commented:
I know, you said you're thinking of it. I'm saying "don't".
Also, please do not merge this until the snap is tested on Ubuntu 20.04 and 22.04, not just 23.10. 
We have already had the sublime-text snap break, and we should be cautious about changing the entire build process.

I'm also uneasy that you've changed the build process in among a process to add workflow. They are separate things, and should be separate PRs.

soumyaDghosh commented:
> I'm also uneasy that you've changed the build process in among a process to add workflow. They are separate things, and should be separate PRs.

I made the `source` of this snap the same as `sublime-text`

https://github.com/snapcrafters/sublime-text/blob/candidate/snap/snapcraft.yaml#L37

Previously `tar` file was used, and just changed it to the deb file. And also added a support for the `dev` builds. There is nothing I changed that is probably dangerous/breaking.

popey commented:
> They are separate things, and should be separate PRs.

soumyaDghosh committed:
revert the change of manifest

jnsgruk commented:
In the spirit of keeping this lean and easy to review; can we also drop the change that introduces `dev` versions? Let's get this bought up to date with the latest workflows that release to candidate first, then we discuss options for `dev` versions later?

jnsgruk commented on .github/workflows/release-to-candidate.yml:
```diff
@@ -0,0 +1,84 @@
+    shell: bash
+    id: check-version
+    run: |
+      version=$(yq -r '.version' snap/snapcraft.yaml | cut -d '-' -f2)
```
Per my latest comment on the PR, this section feels problematic in a workflow that only triggers on pushes to the `candidate` branch. I think we should split this out and discuss separately.

soumyaDghosh replied:
Okay.. I am moving it too.

jnsgruk commented on .github/workflows/sync-upstream.yml:
```diff
@@ -0,0 +1,33 @@
+        with:
+          token: ${{ secrets.SNAPCRAFTERS_BOT_COMMIT }}
+          update-script: |
+            version_details=$(python3 get_latest.py)
```
Feels a bit heavyweight to introduce python here (with a non-std lib requirement on `requests`, no less). Can we rework this to just use shell?

jnsgruk replied:
For example:

```
$ curl -sL https://www.sublimemerge.com/download | grep -Po '<p class="latest"><i>Version:</i> Build \K[0-9]+'
2091

$ curl -sL https://www.sublimemerge.com/dev | grep -Po '<p class="latest"><i>Version:</i> Build \K[0-9]+'
2092
```

soumyaDghosh replied:
But, we'll not able to compare them, like for eg. the `dev` version for `sublime-text` is `4168`  but stable one is `4169`, this one has the `dev` channel a bigger number for now, but how much can we rely on that. If we don't plan at all to support `dev` versions, (There was an issue about that in the sublime-text repo to add dev versions), then yes, definitely yes to shell.

soumyaDghosh replied:
http://mywiki.wooledge.org/ArithmeticExpression

this might come handy though.

soumyaDghosh replied:
Done!

soumyaDghosh committed:
remove the support for dev and need of python file

jnsgruk commented on .github/workflows/sync-upstream.yml:
```diff
@@ -0,0 +1,35 @@
+            # else
+            #     yq -i ".version = \"$dev_version-dev\"" snap/snapcraft.yaml
+            # fi
+            
```
Would be good to tidy up the whitespace here and just end the file with a single newline

soumyaDghosh replied:
sure

jnsgruk commented on .github/workflows/sync-upstream.yml:
```diff
@@ -0,0 +1,35 @@
+        with:
+          token: ${{ secrets.SNAPCRAFTERS_BOT_COMMIT }}
+          update-script: |
+            stable_version=$(curl -sL https://www.sublimemerge.com/download | grep -Po '<p class="latest"><i>Version:</i> Build \K[0-9]+')
```
```suggestion
            stable_version="$(curl -sL https://www.sublimemerge.com/download | grep -Po '<p class="latest"><i>Version:</i> Build \K[0-9]+')"
```

jnsgruk commented on .github/workflows/sync-upstream.yml:
```diff
@@ -0,0 +1,35 @@
+          token: ${{ secrets.SNAPCRAFTERS_BOT_COMMIT }}
+          update-script: |
+            stable_version=$(curl -sL https://www.sublimemerge.com/download | grep -Po '<p class="latest"><i>Version:</i> Build \K[0-9]+')
+            yq -i ".version = \"$stable_version\"" snap/snapcraft.yaml
```
I'd recommend against using `yq` here, as it will reformat a lot of the yaml file.

```suggestion
            sed -i 's/^\(version: \).*$/\1'"$stable_version"'/' snap/snapcraft.yaml
```

soumyaDghosh replied:
I saw that thing happen in the ffmpeg workflow today, so, should I replace all the yqs (in other repos) with sed?

soumyaDghosh replied:
?

jnsgruk commented on snap/snapcraft.yaml:
```diff
@@ -68,4 +68,4 @@ parts:
       - libglu1-mesa
     build-attributes:
-      - no-patchelf
+      - no-patchelf
```
Can you add a newline on the end of this file please?

soumyaDghosh replied:
Sure.. doing all the formatting

jnsgruk reviewed (changes requested):
Some minor changes, but looking much more manageable :)

soumyaDghosh committed:
whitespace cleanups

jnsgruk commented on .github/workflows/sync-upstream.yml:
```diff
@@ -0,0 +1,34 @@
+            # else
+            #     yq -i ".version = \"$dev_version-dev\"" snap/snapcraft.yaml
+            # fi
+            
```
```suggestion

```

soumyaDghosh committed:
remove the last whitespace too

Co-authored-by: Jon Seager <jon@sgrs.uk>

jnsgruk reviewed (dismissed):
LGTM, thanks

jnsgruk commented:
Ah, one last thing before merging - please remove `.github/workflows/test-snap-can-build.yml` - we don't need two workflows to build on PRs :)

soumyaDghosh committed:
Delete .github/workflows/test-snap-can-build.yml as duplicate

soumyaDghosh commented:
@jnsgruk Just wait a minute before merging, I am testing the `source-change`. After that merge both together, as this workflow can't work in this current scenario

soumyaDghosh commented:
Ahh.. merged it.. okay coming up with the second PR
